source/portable/mbedtls/core_pkcs11_mbedtls.c

```diff
@@ -4718,24 +4714,49 @@ CK_DECLARE_FUNCTION( CK_RV, C_GenerateKeyPair )( CK_SESSION_HANDLE hSession,
 
     if( ( xPalPublic != CK_INVALID_HANDLE ) && ( xPalPrivate != CK_INVALID_HANDLE ) )
     {
-        xResult = prvAddObjectToList( xPalPrivate, phPrivateKey, pxPrivateLabel->pValue, pxPrivateLabel->ulValueLen );
+        xAddObjectListResult = prvAddObjectToList( xPalPrivate, phPrivateKey, pxPrivateLabel->pValue, pxPrivateLabel->ulValueLen );
 
-        if( xResult == CKR_OK )
+        if( xAddObjectListResult == CKR_OK )
+        {
+            xAddObjectListResult = prvAddObjectToList( xPalPublic, phPublicKey, pxPublicLabel->pValue, pxPublicLabel->ulValueLen );
+        }
+
+        if( xAddObjectListResult != CKR_OK )
         {
-            xResult = prvAddObjectToList( xPalPublic, phPublicKey, pxPublicLabel->pValue, pxPublicLabel->ulValueLen );
+            LogError( ( "Could not add private key to object list failed with (0x%0lX). Cleaning up PAL objects.", xResult ) );
+
+            xResult = PKCS11_PAL_DestroyObject( xPalPrivate );
```

Comment:
nit: the destroy/delete from list pattern appears at least 3 times in these changes, in a future revision you might make a static method to handle this, i.e. `prvDestroyObjectAndDeleteFromList()` to simplify the code.

Reply:
Noted, will follow up on this in a PR